transformers4rec/torch/model/prediction_task.py

```diff
@@ -417,50 +437,250 @@ class _NextItemPredictionTask(torch.nn.Module):
         softmax_temperature: float
             Softmax temperature, used to reduce model overconfidence, so that softmax(logits / T).
             Value 1.0 reduces to regular softmax.
+        sampled_softmax: Optional[bool]
+            Enables sampled softmax. By default False
+        max_n_samples: Optional[int]
+            Number of samples for sampled softmax. By default 100
+        min_id : Optional[int]
+            The minimum value of the range for the log-uniform sampling. By default 0.
     """
 
     def __init__(
         self,
-        input_size: int,
+        input_size: Sequence,
         target_dim: int,
         weight_tying: bool = False,
         item_embedding_table: Optional[torch.nn.Module] = None,
         softmax_temperature: float = 0,
+        sampled_softmax: Optional[bool] = False,
+        max_n_samples: Optional[int] = 100,
+        min_id: Optional[int] = 0,
     ):
         super().__init__()
         self.input_size = input_size
         self.target_dim = target_dim
         self.weight_tying = weight_tying
         self.item_embedding_table = item_embedding_table
         self.softmax_temperature = softmax_temperature
-        self.log_softmax = torch.nn.LogSoftmax(dim=-1)
+        self.sampled_softmax = sampled_softmax
+
+        if not self.weight_tying:
+            self.output_layer = torch.nn.Parameter(torch.empty(input_size[-1], self.target_dim))
+            torch.nn.init.kaiming_uniform_(self.output_layer, a=sqrt(5))
+
+        if self.sampled_softmax:
+            self.sampler = LogUniformSampler(
+                max_n_samples=max_n_samples,
+                max_id=target_dim,
+                min_id=min_id,
+                unique_sampling=True,
+            )
 
+    def forward(
+        self,
+        inputs: torch.Tensor,
+        targets: Optional[torch.Tensor] = None,
+        training=False,
+        testing=False,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
         if self.weight_tying:
-            self.output_layer_bias = torch.nn.Parameter(torch.Tensor(self.target_dim))
-            torch.nn.init.zeros_(self.output_layer_bias)
+            output_weights = self.item_embedding_table.weight.t()
         else:
-            self.output_layer = torch.nn.Linear(
-                self.input_size[-1], self.target_dim  # type: ignore
-            )
+            output_weights = self.output_layer
+
+        if self.sampled_softmax and (training or testing):
+            logits, targets = self.sampled(inputs, targets, output_weights)
 
-    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
-        if self.weight_tying:
-            logits = torch.nn.functional.linear(
-                inputs,
-                weight=self.item_embedding_table.weight,  # type: ignore
-                bias=self.output_layer_bias,
-            )
         else:
-            logits = self.output_layer(inputs)
+            logits = inputs @ output_weights
```

Comment:
I removed the bias because it think would not be available if ANN is used later for serving. Does that make sense?
I might do some benchmark to see if bias helps to improve accuracy.

Reply:
You're right, that makes sense! Otherwise, we'll need to save the `output_bias` vector in addition to the pre-trained candidate embeddings..